File: dotnet-desktop-guide/winforms/advanced/snippets/how-to-add-data-to-the-clipboard/framework/csharp/ClipboardOperations.cs
using System;
using System.Windows.Forms;

public class Form1 : Form
{
    [STAThread]
    public static void Main()
    {
        Application.EnableVisualStyles();
        Application.SetCompatibleTextRenderingDefault(false);
        Application.Run(new Form1());
    }

    //<CustomerClass>
    [Serializable]
    public class Customer
    {
        private string nameValue = string.Empty;
        public Customer(String name)
        {
            nameValue = name;
        }
        public string Name
        {
            get { return nameValue; }
            set { nameValue = value; }
        }
    }
    //</CustomerClass>

    public Form1()
    {
        //<ClipboardClear>
        Clipboard.Clear();
        //</ClipboardClear>
    }

    //<CustomFormatExample>
    // Demonstrates SetData, ContainsData, and GetData
    // using a custom format name and a business object.
    public Customer TestCustomFormat
    {
        get
        {
            Clipboard.SetData("CustomerFormat", new Customer("Customer Name"));
            if (Clipboard.ContainsData("CustomerFormat"))
            {
                return Clipboard.GetData("CustomerFormat") as Customer;
            }
            return null;
        }
    }
    //</CustomFormatExample>

    //<MultipleFormatsExample>
    // Demonstrates how to use a DataObject to add
    // data to the Clipboard in multiple formats.
    public void TestClipboardMultipleFormats()
    {
        DataObject data = new DataObject();

        // Add a Customer object using the type as the format.
        data.SetData(new Customer("Customer as Customer object"));

        // Add a ListViewItem object using a custom format name.
        data.SetData("CustomFormat",
            new ListViewItem("Customer as ListViewItem"));

        Clipboard.SetDataObject(data);
        DataObject retrievedData = (DataObject)Clipboard.GetDataObject();

        if (retrievedData.GetDataPresent("CustomFormat"))
        {
            ListViewItem item =
                retrievedData.GetData("CustomFormat") as ListViewItem;
            if (item != null)
            {
                MessageBox.Show(item.Text);
            }
        }

        if (retrievedData.GetDataPresent(typeof(Customer)))
        {
            Customer customer =
                retrievedData.GetData(typeof(Customer)) as Customer;
            if (customer != null)
            {
                MessageBox.Show(customer.Name);
            }
        }
    }
    //</MultipleFormatsExample>

    //<GenericSetDataExample>
    // Demonstrates SetData, ContainsData, and GetData.
    public Object SwapClipboardFormattedData(String format, Object data)
    {
        Object returnObject = null;
        if (Clipboard.ContainsData(format))
        {
            returnObject = Clipboard.GetData(format);
            Clipboard.SetData(format, data);
        }
        return returnObject;
    }
    //</GenericSetDataExample>

    //<SetTextExample>
    //<SetAudioExample>
    // Demonstrates SetAudio, ContainsAudio, and GetAudioStream.
    public System.IO.Stream SwapClipboardAudio(
        System.IO.Stream replacementAudioStream)
    {
        System.IO.Stream returnAudioStream = null;
        if (Clipboard.ContainsAudio())
        {
            returnAudioStream = Clipboard.GetAudioStream();
            Clipboard.SetAudio(replacementAudioStream);
        }
        return returnAudioStream;
    }
    //</SetAudioExample>

    //<SetFileDropListExample>
    // Demonstrates SetFileDropList, ContainsFileDroList, and GetFileDropList
    public System.Collections.Specialized.StringCollection
        SwapClipboardFileDropList(
        System.Collections.Specialized.StringCollection replacementList)
    {
        System.Collections.Specialized.StringCollection returnList = null;
        if (Clipboard.ContainsFileDropList())
        {
            returnList = Clipboard.GetFileDropList();
            Clipboard.SetFileDropList(replacementList);
        }
        return returnList;
    }
    //</SetFileDropListExample>

    //<SetImageExample>
    // Demonstrates SetImage, ContainsImage, and GetImage.
    public System.Drawing.Image SwapClipboardImage(
        System.Drawing.Image replacementImage)
    {
        System.Drawing.Image returnImage = null;
        if (Clipboard.ContainsImage())
        {
            returnImage = Clipboard.GetImage();
            Clipboard.SetImage(replacementImage);
        }
        return returnImage;
    }
    //</SetImageExample>

    //<SetHtmlTextExample>
    // Demonstrates SetText, ContainsText, and GetText.
    public String SwapClipboardHtmlText(String replacementHtmlText)
    {
        String returnHtmlText = null;
        if (Clipboard.ContainsText(TextDataFormat.Html))
        {
            returnHtmlText = Clipboard.GetText(TextDataFormat.Html);
            Clipboard.SetText(replacementHtmlText, TextDataFormat.Html);
        }
        return returnHtmlText;
    }
    //</SetHtmlTextExample>
    //</SetTextExample>
}

File: dotnet-desktop-guide/winforms/advanced/snippets/how-to-add-data-to-the-clipboard/framework/vb/ClipboardOperations.vb
Imports System.Windows.Forms

Module Program
    <STAThread>
    Sub Main()
        Application.EnableVisualStyles()
        Application.SetCompatibleTextRenderingDefault(False)
        Application.Run(New Form1())
    End Sub
End Module

Public Class Form1
    Inherits Form

    '<CustomerClass>
    <Serializable()> Public Class Customer

        Private nameValue As String = String.Empty

        Public Sub New(ByVal name As String)
            nameValue = name
        End Sub

        Public Property Name() As String
            Get
                Return nameValue
            End Get
            Set(ByVal value As String)
                nameValue = value
            End Set
        End Property

    End Class
    '</CustomerClass>

    Public Sub New()
        '<ClipboardClear>
        Clipboard.Clear()
        '</ClipboardClear>
    End Sub

    '<CustomFormatExample>
    ' Demonstrates SetData, ContainsData, and GetData
    ' using a custom format name and a business object.
    Public ReadOnly Property TestCustomFormat() As Customer
        Get
            Clipboard.SetData("CustomerFormat", New Customer("Customer Name"))

            If Clipboard.ContainsData("CustomerFormat") Then
                Return CType(Clipboard.GetData("CustomerFormat"), Customer)
            End If

            Return Nothing
        End Get
    End Property
    '</CustomFormatExample>

    '<MultipleFormatsExample>
    ' Demonstrates how to use a DataObject to add
    ' data to the Clipboard in multiple formats.
    Public Sub TestClipboardMultipleFormats()

        Dim data As New DataObject()

        ' Add a Customer object using the type as the format.
        data.SetData(New Customer("Customer as Customer object"))

        ' Add a ListViewItem object using a custom format name.
        data.SetData("CustomFormat", _
            New ListViewItem("Customer as ListViewItem"))

        Clipboard.SetDataObject(data)
        Dim retrievedData As DataObject = _
            CType(Clipboard.GetDataObject(), DataObject)

        If (retrievedData.GetDataPresent("CustomFormat")) Then

            Dim item As ListViewItem = _
                TryCast(retrievedData.GetData("CustomFormat"), ListViewItem)

            If item IsNot Nothing Then
                MessageBox.Show(item.Text)
            End If

        End If

        If retrievedData.GetDataPresent(GetType(Customer)) Then

            Dim customer As Customer = _
                CType(retrievedData.GetData(GetType(Customer)), Customer)

            If customer IsNot Nothing Then

                MessageBox.Show(customer.Name)
            End If

        End If

    End Sub
    '</MultipleFormatsExample>

    '<GenericSetDataExample>
    ' Demonstrates SetData, ContainsData, and GetData.
    Public Function SwapClipboardFormattedData( _
        ByVal format As String, ByVal data As Object) As Object

        Dim returnObject As Object = Nothing

        If (Clipboard.ContainsData(format)) Then
            returnObject = Clipboard.GetData(format)
            Clipboard.SetData(format, data)
        End If

        Return returnObject

    End Function
    '</GenericSetDataExample>

    '<SetTextExample>
    '<SetAudioExample>
    ' Demonstrates SetAudio, ContainsAudio, and GetAudioStream.
    Public Function SwapClipboardAudio( _
        ByVal replacementAudioStream As System.IO.Stream) _
        As System.IO.Stream

        Dim returnAudioStream As System.IO.Stream = Nothing

        If (Clipboard.ContainsAudio()) Then
            returnAudioStream = Clipboard.GetAudioStream()
            Clipboard.SetAudio(replacementAudioStream)
        End If

        Return returnAudioStream

    End Function
    '</SetAudioExample>

    '<SetFileDropListExample>
    ' Demonstrates SetFileDropList, ContainsFileDroList, and GetFileDropList
    Public Function SwapClipboardFileDropList(ByVal replacementList _
        As System.Collections.Specialized.StringCollection) _
        As System.Collections.Specialized.StringCollection

        Dim returnList As System.Collections.Specialized.StringCollection _
            = Nothing

        If Clipboard.ContainsFileDropList() Then

            returnList = Clipboard.GetFileDropList()
            Clipboard.SetFileDropList(replacementList)
        End If

        Return returnList

    End Function
    '</SetFileDropListExample>

    '<SetImageExample>
    ' Demonstrates SetImage, ContainsImage, and GetImage.
    Public Function SwapClipboardImage( _
        ByVal replacementImage As System.Drawing.Image) _
        As System.Drawing.Image

        Dim returnImage As System.Drawing.Image = Nothing

        If Clipboard.ContainsImage() Then
            returnImage = Clipboard.GetImage()
            Clipboard.SetImage(replacementImage)
        End If

        Return returnImage
    End Function
    '</SetImageExample>

    '<SetHtmlTextExample>
    ' Demonstrates SetText, ContainsText, and GetText.
    Public Function SwapClipboardHtmlText( _
        ByVal replacementHtmlText As String) As String

        Dim returnHtmlText As String = Nothing

        If (Clipboard.ContainsText(TextDataFormat.Html)) Then
            returnHtmlText = Clipboard.GetText(TextDataFormat.Html)
            Clipboard.SetText(replacementHtmlText, TextDataFormat.Html)
        End If

        Return returnHtmlText

    End Function
    '</SetHtmlTextExample>
    '</SetTextExample>

End Class